Chapter5/event_manager.erl

```diff
@@ -0,0 +1,89 @@
+-module(event_manager).
+-export([start/2, stop/1]).
+-export([add_handler/3, delete_handler/2, get_data/2, send_event/2, swap_handlers/4]).
+-export([init/1]).
+
+start(Name, HandlerList) ->
+	register(Name, spawn(event_manager, init, [HandlerList])),
+	ok.
+
+init(HandlerList) -> loop( initialize(HandlerList) ).
+
+initialize([]) -> [];
+initialize([{Handler, InitData}|Rest]) ->
+	[ { Handler, Handler:init(InitData) } | initialize(Rest) ].
+
+
+stop(Name) ->
+	Name ! {stop, self()},
+	receive
+		{reply, Reply} -> Reply
+	end.
+
+terminate([]) -> [];
+terminate([{Handler, Data}|Rest]) ->
+	[ { Handler, Handler:terminate(Data) } | terminate(Rest) ] .
+
+
+
+%%Exercise 5-3: Swapping Handlers (Page 162)
+swap_handlers(Name, OldHandler, NewHandler, InitData) ->
+	case add_handler(Name, NewHandler, InitData) of
```

Comment:
Right, all messages will be processed, though chances are that once new handler added and old is not yet removed, both handlers start to consume messages, which is something that we would like to avoid.
I suggest to make a swap operation in event manager context (process), so that it doesn't receive new messages in a process loop until handlers are swapped.